---
title: "Example code for single-cell analysis with Seurat, day 2"
author: "Devika Agarwal, updated by Carla Cohen"
date: "25/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load packages
library(Seurat)
library(tidyverse)
library(patchwork)
library(DT)
library(gprofiler2)
```

# Exercise

## Read in the RDS object we created and save from Seurat day 1

-   use the `readRDS()` function to read in previously saved object

```{r}
seurat_after_qc <- readRDS("seurat_after_qc.rds") # Symlink to RDS file in project directory
```

-   Check the Default Assay

```{r}
DefaultAssay(seurat_after_qc)
```

## Apply SCTransfrom normalisation

Use `SCTransform()` function

SCTransform vignette: <https://satijalab.org/seurat/articles/sctransform_vignette.html>

`SCTransform()` command replaces `NormalizeData()`, `ScaleData` and `FindVariableFeatures()` run for the RNA assay in day 1 Seurat

Should we remove any confounding variables like we did for the RNA assay for Day 1?

Do we want to use the same number of variable features (n=2000) or more than what we used for `NormalizeData()` function.

```{r include = FALSE}
seurat_after_qc <- SCTransform(
  seurat_after_qc,
  variable.features.n = 2000,
  vars.to.regress = "percent_mt"
  )
```

Where is the new normalisation stored?

> Answer: New assay called SCT (required as some genes are filtered out during

Explore the `seurat_after_qc` objects metadata and assays

```{r}
DefaultAssay(seurat_after_qc)
seurat_after_qc[[]] # Access cell metadata
```

Is there a change?

> Answer: Default assay is now SCT

Are there new columns in the metadata?

> Answer: `nCount_SCT` and `nFeature_SCT`

# Exercise

## Visualisation

-   The library size and number of features detected per cell is already present in the Seurat object.

-   When you run `SCTransform()` you get two new variables for library size and features for SCT normalisation

-   Use the function `VlnPlot()` to compare

(i) RNA assay vs SCT assay library size in a single plot
(ii) features detected in RNA vs SCT a single plot

-   Check to see how you can have fixed y axes (ylims) in the `VlnPlot()` function

```{r fig.width = 3, fig.height = 4}
plot1 <- VlnPlot(seurat_after_qc,
        features = c("nCount_RNA", "nCount_SCT"),
        group.by = "orig.ident",
        same.y.lims = TRUE
)

plot2 <- VlnPlot(seurat_after_qc,
        features = c("nFeature_RNA", "nFeature_SCT"),
        group.by = "orig.ident",
        same.y.lims = TRUE
)

cowplot::plot_grid(plot1, plot2, ncol = 1)
```

## Bonus

-   Visualise the distribution of SCT corrected count data for a feature of your choice.
-   Visualise the distribution of SCT logs counts in the data slot for a feature of choice
-   Visualise the distribution of SCT scale.data count feature of your choice.

Let's choose LYZ like day 1

```{r fig.width = 5, fig.height = 1.5}
ggplot_lyz_raw <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "counts"), aes(x = LYZ)) +
    geom_histogram(fill = "grey", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    ggtitle("Raw") +
    cowplot::theme_cowplot()
ggplot_lyz_normalised <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "data"), aes(x = LYZ)) +
    geom_histogram(fill = "grey", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    ggtitle("Log-normalised") +
    cowplot::theme_cowplot()
ggplot_lyz_scaled <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "scale.data"), aes(x = LYZ)) +
    geom_histogram(fill = "grey", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    ggtitle("Scaled") +
    cowplot::theme_cowplot()


cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised,ggplot_lyz_scaled, ncol = 3)
```

Use the function `VariableFeatures` to pull out the 1:10 the variable genes after SCT and compare to 1:10 from the RNA assay

Do we need to change any arguments to get the variables genes specific to the SCT or RNA assay

```{r}
rna_vf <- VariableFeatures(seurat_after_qc, assay = "RNA")[1:10]
sct_vf <- VariableFeatures(seurat_after_qc, assay = "SCT")[1:10]
```

How do the two gene lists compare?

```{r}
print("Intersection: ")
intersect(rna_vf, sct_vf)
print("Difference: ")
setdiff(rna_vf, sct_vf)
```

# Exercise

## Dimensionality reduction on SCT transformed data

-   Run a principal component analysis and UMAP on the Seurat object.

-   Check the Default assay

-   Do we want to change the `reduction.name` argument so that we can still keep the RNA assay based PCA results?

```{r}
seurat_after_qc <- RunPCA(seurat_after_qc, assay = "SCT", reduction.name = "pca.sct")
```

Check to see what reductions are now present in the object

```{r}
Reductions(seurat_after_qc)
```

-   First, visualise the amount of variance explained the top principal components for SCTransformed data (number of your choice). How many principal components would you use for downstream analyses?

-   Do we need to specify the reduction?

```{r}
ElbowPlot(seurat_after_qc, ndims = 50, reduction = "pca.sct")
```

-   Run the UMAP technique on the first 20 principal components stored in the Seurat object and visualise the result as a scatterplot.

How can we change the reduction name from default "umap" to "sct.umap"

How can we specify that we want to use PCA run on the SCT Assay (sct.pca) in the previous step?

```{r include = FALSE}
seurat_after_qc <- RunUMAP(seurat_after_qc, assay = "SCT", dims = 1:20, reduction = "pca.sct", reduction.name = "umap.sct", reduction.key = "SCTUMAP_")
```

Use DimPlot() to plot the umap. What happens if you try to specify different reductions with UMAPPlot?

Compare RNA based umap with sct.umap

```{r fig.width = 5, fig.height = 2}
p1 <- DimPlot(seurat_after_qc, reduction = "umap", group.by = "RNA_snn_res.0.5") + ggtitle("Default UMAP")
p2 <- DimPlot(seurat_after_qc, reduction = "umap.sct", group.by = "orig.ident") + ggtitle("SCT UMAP")
p1 + p2
```

# Exercise

## Clustering on SCTransformed data

-   Compute the graph of nearest neighbours using the function `FindNeighbors()`. Which principal components are used by default? Instead, specify the number of components that you have chosen. Have you chosen the right `reduction` and `assay`?

```{r include = FALSE}
seurat_after_qc <-  FindNeighbors(seurat_after_qc,
                                  assay = "SCT",
                                  reduction = "pca.sct",
                                  dims = 1:20,
                                  k.param = 20
                                  )
```

-   Finally, compute cluster labels. What is the default setting for the `resolution` argument? Instead, set it to `0.5`. Do you expect more or fewer clusters following that change? What other parameters would you also try to experiment with?
    -   should we specify the `graph.name`?

```{r include = FALSE}
seurat_after_qc <- FindClusters(seurat_after_qc,
                                graph.name = "SCT_snn",
                                resolution = 0.5,
                                algorithm = 1
                                )
```

Check cluster assignment between SCT and RNA workflow

If you use the same resolution = o.5 and dims as RNA workflow do you get the same number of cluster or more?

Are cells in the same cluster across both RNA and SCT

```{r}
table(seurat_after_qc$RNA_snn_res.0.5, seurat_after_qc$SCT_snn_res.0.5)
```

-   Visualise the SCT cluster labels on the SCT transformed UMAP scatter plot and the RNA cluster labels on the RNA UMAP How would you describe the agreement between the UMAP layout and clustering for SCT vs RNA Assay results?

```{r fig.width = 5, fig.height = 2}
p3 <- DimPlot(seurat_after_qc, reduction = "umap.sct", group.by = "SCT_snn_res.0.5") + ggtitle("SCT UMAP")
p1 + p3
```

Plot some known cell-type markers for PBMC datasets, does the SCT better separate the celltypes?

CD14+ Monocyte : LYZ, CD14 CD16 Monocytes : FCGR3A, MS4A7 CD4 T : CD4, IL76 CD8 T : CD8A, CD3D NK : GNLY, GZMB,NKG7 B Cell : MS4A1 , CD79A DC : CST3, FCER1A Platelets : PPBP

```{r, fig.width = 10, fig.height = 4}
FeaturePlot(seurat_after_qc, features = c("LYZ","FCGR3A","CD4","CD8A","GNLY","MS4A1","CST3","PPBP"), reduction = "umap", ncol = 4)
FeaturePlot(seurat_after_qc, features = c("LYZ","FCGR3A","CD4","CD8A","GNLY","MS4A1","CST3","PPBP"), reduction = "umap.sct", ncol = 4)
```

Calculate the markers for these clusters from either the RNA or SCT assay

```{r include = FALSE}
# Idents(seurat_after_qc) <- "RNA_snn_res.0.5"
# seurat_markers_all <- FindAllMarkers(seurat_after_qc,
#                                      assay = "RNA",
#                                      logfc.threshold = 0.25,
#                                      min.pct = 0.25,
#                                      test.use = "wilcox"
#                                      )

Idents(seurat_after_qc) <- "SCT_snn_res.0.5"
seurat_markers_all_sct <- FindAllMarkers(seurat_after_qc,
                                         assay = "RNA",
                                         logfc.threshold = 0.25,
                                         min.pct = 0.25,
                                         test.use = "wilcox")
```

# Bonus exercise to try in your own time: Pathway analysis on Cluster markers for all clusters

-   Choose either RNA workflow based generated or SCT generated based Seurat marker results

-   we will be using Gprofiler `gost()` function for multiple gene lists at the same time

-   First we need to filter Seurat top significant (p_adj \< 0.05) upregulated genes with a LogFC threshold (decided by you) for each cluster, use `dplyr:: group_by()` and `dplyr::filter` to get gene list for each cluster, then only select cluster and gene

-   We then use `split()` on the filtered_df to divide gene markers into list of multiple character vectors containing genes split by cluster

-   can refer to the pathway analysis code from previous tutorial, but use human not mouse pathways

First generate the list of markers for each cluster

```{r}
seurat_markers_all_sct
```

```{r}
seurat_clusters_results_filtered <- seurat_markers_all_sct %>% 
  group_by(cluster) %>% 
  filter(p_val_adj < 0.05, avg_log2FC >= 1) %>% 
  select(cluster, gene)

seurat_clusters_list <- split(seurat_clusters_results_filtered$gene, seurat_clusters_results_filtered$cluster)
```

```{r}
seurat_clusters_list
```

We then run pathway analysis using `gost()` with multi_query = TRUE

-   To generate all_genes_id , we use all genes present in either the RNA assay or SCT assay, as we have already filtered out lowly expressed genes not present in certain cells.

```{r}
# Choose Default assay based on if running pathway analyses on RNA or SCT results
DefaultAssay(seurat_after_qc) <- "SCT"
# create a vector of  of all genes 
all_genes_id <- rownames(seurat_after_qc)


multi_gostquery_results_obj <- gost(query = seurat_clusters_list,
                                    organism = "hsapiens",
                                    multi_query = TRUE,
                                    domain_scope = "custom",
                                    custom_bg = all_genes_id
                                    )
```

can you plot the results for different clusters together ?

```{r fig.width = 10, fig.height = 10}
gostplot(multi_gostquery_results_obj)
```

## Afternoon Session

Demultiplexing with hashtag oligos (HTOs)

Dataset : 12-HTO dataset from four human cell lines

-   Dataset represent Data represent single cells collected from four cell lines: HEK, K562, KG1 and THP1
-   Each cell line was further split into three samples (12 samples in total)
-   Each sample was labeled with a hashing antibody mixture (CD29 and CD45), pooled, and run on a single lane of 10X.
-   Based on this design, we should be able to detect doublets both across and within cell types

Load in the UMI matrix for the RNA data and check the dimensions

```{r}
hto12.umis <- readRDS("hto12_umi_mtx.rds") # Symlink to RDS file in project directory
```

```{r}
dim(hto12.umis)
```

What do rows and columns represent?

> Answer: Rows represent features (genes) and columns represent cell barcodes (non-empty droplets called by cellranger pipeline)

Load in the HTO matrix and check the dimensions - What is the structure of the hto12.htos object? (what is the rows and what is the columns?) - Check the class of the HTO object - How many cell barcodes and HTO samples do we have?

```{r}
hto12.htos <- readRDS("hto12_hto_mtx.rds") # Symlink to RDS file in project directory
```

```{r}
dim(hto12.htos)
```

```{r}
head(hto12.htos)
```

Now we only want to subset to those cell barcodes or cells (actually called as cells by cellRanger or EmptyDrops on the gene expression data) which are detected by both RNA and HTO matrices - Check the class of the joint.bcs object and how many cell barcodes do we have in common

```{r}
joint.bcs <- intersect(colnames(hto12.umis), rownames(hto12.htos))
```

Subset the RNA matrix to only the `joint.bcs` cell barcodes and check the dimensions

```{r}
hto12.umis.common <- hto12.umis[, joint.bcs]
dim(hto12.umis.common)
```

-   Create a Seurat object with the RNA (UMI count matrix) data using only the joint bcs

Name the object `hto12_object` Include features detected in at least 3 cells, and cells where at least 200 features detected Normalise with log normalsiation ,find variable genes and Scale RNA data

```{r include = FALSE}
hto12_object <- CreateSeuratObject(
  counts = hto12.umis.common,
  project = "hto12",
  assay = "RNA", 
  min.cells = 3,
  min.features = 200
)
hto12_object <- NormalizeData(hto12_object, assay = "RNA")
hto12_object <- FindVariableFeatures(hto12_object, assay = "RNA", selection.method = "vst")
hto12_object <- ScaleData(hto12_object, assay = "RNA", features = VariableFeatures(hto12_object))
```

```{r}
dim(hto12_object)
```

Add HTO data as another assay to `hto12_object`

-   First subset the HTO matrix to those cell barcodes which are now in the `hto12_object` Seurat object and make sure that the features only consist of the HTO tags
-   Is our subsetted `hto12.htos.common` in the right format? if not what do we do to get it in the right format before adding it as another assay?

> Answer : Remove additional (non-HTO) columns, convert to matrix and transpose

```{r}
hto12.htos.common <- hto12.htos[colnames(hto12_object), 1:12] %>% as.matrix() %>% t()
dim(hto12.htos.common)
```

Now use `CreateAssayObject()` to add the subsetted HTO matrix to the already created `hto12_object` seurat object as a new assay called `HTO`

```{r include = FALSE}
hto12_object[["HTO"]] <- CreateAssayObject(counts = hto12.htos.common)
```

Normalise the HTO data , here we will use the CLR transformation with margin =1 (Default setting) CLR: Applies a centered log ratio transformation This is required because the HTO data is bimodal, i.e. the tag is either present or absent on the cell. This is quite different to the RNA counts.

```{r include = FALSE}
hto12_object <- NormalizeData(hto12_object, assay = "HTO", normalization.method = "CLR" , margin = 1) # Margin 1 specifies to normalise across features (only required for CLR normalisation)
```

Demultiplex cells based on HTO enrichment

Here we use Seurat Function `HTODemux()` to assign single cells to their original samples

```{r include = FALSE}
hto12_object <- HTODemux(hto12_object)
```

Checkout the metadata column of the `hto12_object`, try to read the `HTODemux()` results output summary in the `Value` section to understand the results

```{r}
hto12_object[[]]
```

Visualise the Demultiplexing results

We can visualise how many cells are classified as singlets, doublets and negative/ambiguous cells

Check the meta.data, which column do we want for this information?

```{r}
table(hto12_object$HTO_classification.global)
```

Visualize enrichment for selected HTOs with ridge plots

plot the max HTO signal for one of the HTO of each of the 4 cell lines (HEK, K562, KG1 and THP1) features with ridge plots using the `RidgePlot()` function

plot Max HTO signal

```{r, fig.height=8, fig.width=10}
# Change the identities of the seurat object to the relevant metadata column
Idents(hto12_object) <- hto12_object$hash.ID
RidgePlot(hto12_object, features = c("HEK-A", "K562-A", "KG1-A", "THP1-A"), ncol = 2)
```

Visualize pairs of HTO signals to confirm mutual exclusivity in singlets between the same celline

a)  plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by (single/doublet/negative status)

b)  plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_maxID

c)  plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_secondID

use the function `FeatureScatter()`

```{r}
DefaultAssay(hto12_object) <- "HTO"
FeatureScatter(hto12_object, "HEK-A", "HEK-B", group.by = "HTO_classification.global")
FeatureScatter(hto12_object, "HEK-A", "HEK-B", group.by = "HTO_maxID")
FeatureScatter(hto12_object, "HEK-A", "HEK-B", group.by = "HTO_secondID")
```

# Bonus Exercise

Plot scatter plot of 2 HTOs across different cell lines e.g. K562 vs KG1 and colour by (single/doublet/negative status) and HTO_max ID

```{r}
FeatureScatter(hto12_object, "K562-A", "KG1-A", group.by = "HTO_classification.global")
```

Compare number of RNA UMIs for singlets, doublets and negative cells

What is a suitable plot for such comparisons?

> Answer: Violin plot

```{r}
Idents(hto12_object) <- "HTO_classification.global"
VlnPlot(hto12_object, features = "nCount_RNA")
```

Question: what do you notice?

> Answer: Doublets have higher UMI counts on average, but there is significant overlap with singlets; therefore, UMI count is insufficient for doublet detection

Visualize HTO signals in a heatmap , lookup `HTOHeatmap()`

```{r, fig.width=8}
HTOHeatmap(hto12_object)
```

What do you notice?

Generate a two dimensional tSNE or UMAP embedding for HTOs. Here we are grouping cells by singlets and doublets ONLY for simplicity.

Do we need to subset our object?

If so what are we subsetting out?

```{r}
Idents(hto12_object) <- "HTO_classification.global"
hto12_object.subset <- subset(hto12_object, subset = HTO_classification.global != "Negative")
```

# Run UMAP/TSNE

what assay are we running UMAP/tsne for ?

look up the arguments in `RunUMAP()` and/or `RunTSNE()` functions

check which arguments in RunUMAP/RunUMAP/RunTSNE can be used to change the name of the reduction from defauult name of pca/umap/tsne to custom name

before we Run UMAP, we need to scale and run PCA like we did in the normal single cell workflow

> Answer: HTO

```{r include = FALSE}
# Calculate a tSNE & UMAP embedding of the HTO data
DefaultAssay(hto12_object.subset) <- "HTO"
hto12_object.subset <- ScaleData(hto12_object.subset, assay = "HTO", features = rownames(hto12_object.subset))
hto12_object.subset <- RunPCA(hto12_object.subset, assay = "HTO", features = rownames(hto12_object.subset), npcs = 5, reduction.name = "pca.hto", reduction.key = "HTOPC_")
hto12_object.subset <- RunUMAP(hto12_object.subset, assay = "HTO", reduction = "pca.hto", dims = 1:5, reduction.name = "umap.hto", reduction.key = "HTOUMAP_")
```

```{r}
ElbowPlot(hto12_object.subset, ndims = 5, reduction = "pca.hto")
```

check the Reductions in the object

```{r}
Reductions(hto12_object.subset)
```

Plot the UMAP/tsne for the HTO assay \* which reduction shall we plot?

• colour by if singlet/doublet

• colour by HTO final classification results (hash.ID)

-   check the arguments on how label the clusters by the cluster identity

-   can change the label size?

```{r fig.width = 5, fig.height = 2}
p1 <- DimPlot(hto12_object.subset,
        reduction = "umap.hto",
        group.by = "HTO_classification.global"
        )
p2 <- DimPlot(hto12_object.subset,
        reduction = "umap.hto",
        group.by = "hash.ID",
        label = TRUE,
        label.size = 3,
        repel = TRUE
        )
p1 + p2
```

what do you notice about the cluustering on tthe UMAP/tsne, does the number of clusters mean anything?

> Answer: Number of clusters = number of HTOs

what do you notice about the cloud of cells surrounding each cluster?

> Answer: Doublets

## Bonus exercises

You can also visualize the more detailed classification result by group.by HTO_maxID before plotting.

```{r}
DimPlot(hto12_object.subset,
        reduction = "umap.hto",
        group.by = "HTO_maxID",
        label = TRUE,
        label.size = 3,
        repel = TRUE
        )
```

What happens if you group.by the UMAP/TSNE plot by HTO_maxID?

> Answer: Removes doublet classification (assigns doublets to max ID)

Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

do we need to rerun `FindVariableFeatures()` and `ScaleData()` again?

> Answer : Yes (subsetted Seurat object)

what other steps do we need run to get visualise our RNA data as UMAP/t-SNE coloured by doublets/singlets and celltypes?

Answer:

```{r include = FALSE}
DefaultAssay(hto12_object.subset) <- "RNA"
# Run PCA on most variable features
hto12_object.subset <- FindVariableFeatures(hto12_object.subset, assay = "RNA", selection.method = "vst")
hto12_object.subset <- ScaleData(hto12_object.subset, assay = "RNA", features = VariableFeatures(hto12_object.subset))


hto12_object.subset <- RunPCA(hto12_object.subset, assay = "RNA")
hto12_object.subset <- RunUMAP(hto12_object.subset, assay = "RNA", dims = 1:8)
```

Plot RNA based UMAP

group.by hash.ID

create a new seurat object meta.data column called cell_line , which removes "\_A or \_B or \_C " in the hash.ID and replaces it with "", to create a new meta.data with only the cell-line info

```{r}
#we create another metadata column based on the hash.id column, where we gsub the HTO tag info (-A,-B,-C) for each cell line to plot only the cell lien names to see if we have batch effect
hto12_object.subset$cell_line <- gsub(pattern = "[-ABC]", x = hto12_object.subset$hash.ID, replacement = "")
hto12_object.subset[[]]
```

```{r}
Idents(hto12_object.subset) <- "cell_line"
UMAPPlot(hto12_object.subset)
```

what does our RNA based clustering on the UMAP/T-SNE show?

> Answer: Cell lines have distinct gene expression profiles; doublets highlighted

## Bonus exercise (try in your own time)

create a second seurat object based , using the code above, and rerun the `HTODemux()` with a different value of positive quantile.

try to check if the classification changes massively if you adjusted the threshold for classification by playing around with the `positive.quantile` argument from the default.
