---
title: "Example code for single-cell analysis with Seurat, day 1"
author: "Kevin Rue-Albrecht"
date: "05/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load packages
library(Seurat)
library(tidyverse)
library(cowplot)
```

# Exercise

## Import scRNA-seq data and create a Seurat object

-   Use the function `Read10X()` to import data in the directory `filtered_feature_bc_matrix/` as an object named `read10x_data`. What class of object does the function return?

```{r}
read10x_data <- Read10X(
  "filtered_feature_bc_matrix", # Relative path to symlink to data folder in project directory
  strip.suffix = TRUE
)
```

```{r}
class(read10x_data)
```

> Answer: `dgCMatrix` (type of sparse matrix)

-   Have a look at the object and its structure (e.g., first 15 rows and 6 columns). What is a sparse matrix and how does it store data?

```{r}
read10x_data[1:15, 1:6]
```

```{r}
str(read10x_data)
```

> Answer:
>
> Internal slots of `dgCMatrix` class objects:
>
> -   `x` - vector of all non-zero values in column-oriented form
>
> -   `p` - vector of zero-based indices within `x` of first non-zero element for each column
>
> -   `i` - vector of zero-based indices of row numbers (sorted in ascending order) for each element in `x`
>
> -   `Dim`, `DimNames` - matrix dimensions and character vectors of row and column names respectively

-   How many features and barcodes (i.e., cells) are present in the data set?

```{r}
dim(read10x_data)
```

> Answer: 33538 features; 5155 barcodes

-   Create a Seurat object using the function `CreateSeuratObject()` and the object `read10x_data`. Name the object `seurat_object`. Include features detected in at least 3 cells, and cells where at least 200 features detected. Name the project `pbmc5k`. How many features and barcodes are left in the Seurat object?

```{r}
seurat_object <- CreateSeuratObject(
  read10x_data,
  project = "pbmc5k",
  min.cells = 3,
  min.features = 200
  )
seurat_object
```

> Answer: 19037 features; 5100 barcodes

-   How many features and cells were filtered out when you created the Seurat object?

```{r}
dim(read10x_data) - dim(seurat_object)
```

> Answer: 14501 features; 55 barcodes

# Exercise

## Accessing the contents of a Seurat object

-   Query the name of the default assay in the Seurat object.

```{r}
DefaultAssay(seurat_object)
```

-   List the names of assays available in the Seurat object.

```{r}
Assays(seurat_object)
```

-   Display the first six rows and six columns of the RNA assay data. What function do you use? Which arguments do you need to supply, and which ones are optional?

```{r}
GetAssayData(seurat_object)[1:6, 1:6]
```

> Answer: `GetAssayData()`; required = `object`; optional = `slot`, `assay`

-   Display the entire data.frame of per-cell metadata (first six rows). What column names do you see?

```{r}
head(seurat_object@meta.data)
```

> Answer: orig.ident, nCount_RNA, nFeature_RNA

-   Fetch one column of metadata using `[[`. What type of object do you get back?

```{r}
head(seurat_object[["nCount_RNA"]])
```

> Answer: Data frame

-   Instead,fetch the same column of metadata using `$`. What type of object do you get back this time?

```{r}
head(seurat_object$nCount_RNA)
```

> Answer: Vector (named)

-   Use the function `FetchData()` to access the library size and expression of the feature named `"LYZ"` (first six rows). What type of object do you get back?

```{r}
# FetchData() is a very useful function to extract multiple columns of data from a Seurat object as a data frame for use with e.g. ggplot2
FetchData(seurat_object, vars = c("nCount_RNA", "LYZ"))
```

> Answer: Data frame

# Demo

## Common operations on Seurat objects

`WhichCells()` returns the names of cells that match a logical expression.

```{r}
WhichCells(seurat_object, expression = LYZ > 500)
```

`VariableFeatures()` returns the names of variable features (for a given assay, if computed).

```{r}
VariableFeatures(seurat_object)
```

`subset()` returns a new Seurat object restricted to certain features and cells.

```{r}
# Verbose non-tidyverse syntax
subset(
    seurat_object,
    cells = WhichCells(seurat_object, expression = LYZ > 500),
    features = VariableFeatures(seurat_object)
)

# Less verbose tidyverse pipe syntax
seurat_object %>% 
  subset(cells = WhichCells(., expression = LYZ > 500),
         features = VariableFeatures(.)
         )
```

# Exercise

## Quality control and visualisation

-   The library size and number of features detected per cell is already present in the Seurat object. Use the function `VlnPlot()` to display them in a single violin plot.

```{r}
VlnPlot(seurat_object, features = c("nCount_RNA", "nFeature_RNA"))
```

-   Use the function `PercentageFeatureSet()` to compute the fraction of reads assigned to mitochondrial genes in each cell. Store the metric in the cell metadata of the Seurat object, under the name `"percent_mt"`. Visualise this new metric alongside the previous two in a new violin plot.

```{r}
seurat_object$percent_mt <- PercentageFeatureSet(
  seurat_object,
  pattern = "^MT-" # Mitochondrial gene names in this dataset start with 'MT-'
)

VlnPlot(seurat_object, features = c("nCount_RNA", "nFeature_RNA", "percent_mt"))
```

-   Visualise a scatter plot of the proportion of mitochondrial UMIs against the library size in each cell.

```{r}
FeatureScatter(seurat_object, "percent_mt", "nCount_RNA")
```

-   Create a new Seurat object, called `seurat_after_qc`, that is subsetted to cells that have more than 4,500 UMI counts, less than 15% of UMI counts assigned to mitochondrial features, and more than 1,500 features detected. How many cells were removed in this step?

```{r}
seurat_after_qc <- subset(seurat_object, subset = (nCount_RNA > 4500 & percent_mt < 15 & nFeature_RNA > 1500))
seurat_after_qc
```

```{r}
dim(seurat_object) - dim(seurat_after_qc)
```

> Answer: 896

# Exercise

## Normalisation

-   Normalise the RNA assay of the Seurat object (after quality control) using the `"LogNormalize"` method.

```{r include = FALSE}
seurat_after_qc <- NormalizeData(
    seurat_after_qc,
    normalization.method = "LogNormalize"
)
```

## Bonus

-   Visualise the distribution of raw counts and normalised data for a feature of your choice.

```{r}
str(seurat_after_qc)
```

```{r}
FetchData(seurat_after_qc, vars = "LYZ", slot = "data")
```

```{r}
ggplot_lyz_raw <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "counts"), aes(x = LYZ)) +
    geom_histogram(bins = 50) +
    coord_cartesian(ylim = c(0, 500)) +
    labs(x = "LYZ (raw)") +
    cowplot::theme_cowplot()
ggplot_lyz_normalised <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "data"), aes(x = LYZ)) +
    geom_histogram(bins = 50) +
    coord_cartesian(ylim = c(0, 500)) +
    labs(x = "LYZ (log-normalised)") +
    cowplot::theme_cowplot()
cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised, ncol = 1)
```

# Exercise

## Variable features and scaling

-   Identify variable features in the normalised RNA assay of the Seurat object. Use the `"vst"` method and select the 2,000 most variable features. What does this subsetting do, and what are our motivations for doing it?

```{r include = FALSE}
seurat_after_qc <- FindVariableFeatures(
    seurat_after_qc,
    selection.method = "vst",
    nfeatures = 2000
)
```

> Answer: Reduces number of genes used in downstream analysis; required to reduce computational burden of storing scaled (i.e. non-sparse) counts for dimensionality reduction and clustering

-   What is the function to display the name of variable features in a Seurat object (e.g., first 10)? How can you control which assay the variable features are pull from?

```{r}
head(VariableFeatures(seurat_after_qc), n = 10)
```

> Answer: See above; `assay` argument can be used to specify which assay to pull variable features from

-   Use the function `VariableFeaturePlot()` to visualise the scatter plot of standardised variance against average expression. How would you use this plot?

```{r}
VariableFeaturePlot(seurat_after_qc)
```

> Answer: Confirm appropriate number of highly variable genes have been chosen for the dataset

-   Scale the normalised RNA assay of the Seurat object, regressing the library size and the fraction of UMI counts assigned to mitochondrial features. What are the motivations for removing those two sources of variation?

```{r include = FALSE}
seurat_after_qc <- ScaleData(
  seurat_after_qc,
  vars.to.regress = c("nCount_RNA", "percent_mt")
)
```

> Answer: Remove artefactual sources of variation in gene expression

# Exercise

## Dimensionality reduction

-   Run a principal component analysis on the Seurat object. Which features are used by the method in the default settings? How could you change this? How do you read the message output of the function `RunPCA()`?

```{r}
seurat_after_qc <- RunPCA(seurat_after_qc)
```

> Answer: Top 20 genes per PC based on loading scores (positive and negative)

-   List the names of dimensionality reduction results available in the Seurat object.

```{r}
Reductions(seurat_after_qc)
```

-   Use `PCAPlot()` or `DimPlot()` to produce a scatterplot of the first and second PCA components.

```{r}
PCAPlot(seurat_after_qc)
```

## Bonus

-   Make a scatterplot of the first and second PCA components yourself using `ggplot2`.

```{r}
# Use this code chunk to prepare a data.frame for ggplot2
pca_data <- FetchData(seurat_after_qc, vars = c("PC_1", "PC_2"))
head(pca_data)
```

```{r}
ggplot(pca_data, aes(x = PC_1, y = PC_2)) +
  geom_point(size = 0.2) +
  cowplot::theme_cowplot()
```

-   Visualise the amount of variance explained the top principal components (number of your choice). How many principal components would you use for downstream analyses?

```{r}
ElbowPlot(seurat_after_qc, ndims = 50)
```

> Answer: 15-20

-   Run the UMAP technique on your selected number of principal components and visualise the result as a scatterplot.

```{r include = FALSE}
seurat_after_qc <- RunUMAP(seurat_after_qc, dims = 1:20)
```

```{r}
UMAPPlot(seurat_after_qc)
```

# Exercise

## Clustering

-   Compute the graph of nearest neighbours using the function `FindNeighbors()`. Which principal components are used by default? Instead, specify the number of principal components that you have chosen earlier.

```{r include = FALSE}
seurat_after_qc <- FindNeighbors(
  seurat_after_qc,
  dims = 1:20,
  k.param = 20
)
```

> Answer: The help page states that the function `FindNeighbors()` uses principal components 1 through 10, by default.

-   What are the names of the nearest neighbour graphs that are now stored in the Seurat object?

```{r}
Graphs(seurat_after_qc)
```

-   Finally, compute cluster labels. What is the default setting for the `resolution` argument? Instead, set it to `0.5`. Do you expect more or fewer clusters following that change? What other parameters would you also try to experiment with?

```{r}
res <- seq(0.5, 1, 0.1) # Vector of resolution parameters from 0.5 to 1 in 0.1 increments
```

```{r include = FALSE}
seurat_after_qc <- FindClusters(
  seurat_after_qc,
  graph.name = "RNA_snn",
  resolution = res,
  algorithm = 1 # Community detection algorithm (default is Louvain)
)
```

-   Visualise the cluster labels on the UMAP scatter plot. How would you describe the agreement between the UMAP layout and the clustering results?

```{r}
UMAPPlot(
  seurat_after_qc,
  group.by = "RNA_snn_res.0.5",
  label = TRUE
)
```

```{r}
# Generate UMAP plots for each cluster resolution
cluster_plots <- lapply(
  res,
  FUN = function(x) UMAPPlot(seurat_after_qc, group.by = paste0("RNA_snn_res.", x), label = TRUE)
)
```

```{r fig.width = 4, fig.height = 5}
# Arrange UMAP plots for all cluster resolutions in a grid
cowplot::plot_grid(
  plotlist = cluster_plots,
  ncol = 2
)
```

# Exercise

## Identify cluster markers

-   Use the function `FindAllMarkers()` to identify positive markers for all clusters, filtering markers that are detected in at least 25% of the cluster, and with a log fold-change greater than `0.25`. Assign the result to an object named `seurat_markers_all`. What is the class of that object? How do you control the set of clusters that are used?

```{r}
Idents(seurat_after_qc) <- "RNA_snn_res.0.5"
```

```{r include = FALSE}
seurat_markers_all <- FindAllMarkers(
  seurat_after_qc,
  logfc.threshold = 0.25,
  min.pct = 0.25,
  test.use = "wilcox"
)
class(seurat_markers_all)
```

> Answer: Data frame

-   How do you read the contents of the object `seurat_markers_all`? How do you know which features are the markers of each cluster?

```{r}
head(seurat_markers_all)
```

> Answer:

-   Filter and display the top 10 markers for cluster 3.

```{r}
cluster3_top10 <- seurat_markers_all %>% 
  filter(cluster == 3, p_val_adj < 0.05) %>% 
  slice_max(order_by = avg_log2FC, n = 10) %>% 
  pull(gene)
```

-   Visualise the expression of the top 4 marker for cluster 3 on a UMAP layout.

```{r fig.width = 4, fig.height = 4}
FeaturePlot(seurat_after_qc, features = cluster3_top10[1:4], label = TRUE)
```

-   Visualise the expression of those same 4 marker as a violin plot. Do you have any particular preference between the two types of plots?

```{r}
VlnPlot(seurat_after_qc, features = cluster3_top10[1:4], ncol = 2)
```

> Answer: Violin plot clearer to differentiate between clusters but does not show relationship between clusters

-   Use `DoHeatmap()` to visualise the top 10 (positive) markers for each cluster. Hint: you may want to use the function `dplyr::group_by()`.

```{r}
markers_top10_clusters <- seurat_markers_all %>%
  filter(p_val_adj < 0.05) %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n = 10) %>% 
  pull(gene)
```

```{r fig.width = 7, fig.height = 5}
DoHeatmap(
  seurat_after_qc,
  features = markers_top10_clusters,
  group.by = "RNA_snn_res.0.5",
  label = FALSE
)
```
